src/classes/luaCoroutine.cpp

```diff
@@ -226,3 +228,7 @@ void LuaCoroutine::terminate_hook(lua_State* tState, lua_Debug* dbg) {
    lua_pushstring(tState, "execution terminated");
    lua_error(tState);
 }
+
+bool LuaCoroutine::isRunning() {
+    return (!done) && lua_status(tState) == LUA_OK;
+}
```

Comment:
If the intention is for this to be called on other threads, this is not thread safe. We can only access the lua state aka `tState` here from a single thread at a time.

Reply:
The best way to handle this is probably a boolean value set before and after lua resumes.

Reply:
Come to think of it, this may also be an issue with the hooks.

Reply:
I mean, there is only so much the interpreter can do. We aren't modifying the lua interpreter, so the only thing we can do is set some hooks. That way it's set up to be run on the lua thread to be executed. This is kinda the only thing it does. It defers a call to yield (or error, depending on what you look at) with sethook

Reply:
Plus, lua_status only reads, which would technically mean it's thread safe.

Reply:
> Plus, lua_status only reads, which would technically mean it's thread safe.

Thats not how thread safety works though. Just because it only reads dosnt make it thread safe. It could be reading the state while the state is also being modified by another thread. According to lua the only thread safety guarantee is to have a separate lua state on each thread. 

I am working on a solution now.

Reply:
Another issue is the hook effects the parent and sibling states as well.

Reply:
Oh okay, I figured out the solution would be to make a function to register the hook while it's not running yet, and make it check for variables inside the actual coroutine object. But since you are looking at it from a static point of view, that could get it messy real quick.

Reply:
> Oh okay, I figured out the solution would be to make a function to register the hook while it's not running yet, and make it check for variables inside the actual coroutine object. But since you are looking at it from a static point of view, that could get it messy real quick.

That would work, the problem is hooks come with a big performance loss. So I want the user to be in control of if and when hooks are there. I am exposing full hook functionality directly to GDScript is my plan.

Reply:
Looks good to me.